src/main/java/org/candlepin/subscriptions/security/IdentityHeaderAuthenticationFilter.java

```diff
@@ -0,0 +1,93 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.subscriptions.security;
+
+import org.candlepin.subscriptions.exception.ErrorCode;
+import org.candlepin.subscriptions.exception.SubscriptionsException;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.web.authentication.preauth.AbstractPreAuthenticatedProcessingFilter;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedCredentialsNotFoundException;
+import org.springframework.util.StringUtils;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Response;
+
+/**
+ * Spring Security filter responsible for pulling the principal out of the x-rh-identity header
+ */
+public class IdentityHeaderAuthenticationFilter extends AbstractPreAuthenticatedProcessingFilter {
+    private static final Logger log = LoggerFactory.getLogger(IdentityHeaderAuthenticationFilter.class);
+    public static final String RH_IDENTITY_HEADER = "x-rh-identity";
+    private ObjectMapper mapper;
+
+    public IdentityHeaderAuthenticationFilter(ObjectMapper mapper) {
+        this.mapper = mapper;
+    }
+
+    @Override
+    protected Object getPreAuthenticatedPrincipal(HttpServletRequest request) {
+        try {
+            String identityHeader = request.getHeader(RH_IDENTITY_HEADER);
+
+            if (StringUtils.isEmpty(identityHeader)) {
+                log.debug("{} is empty", RH_IDENTITY_HEADER);
+                // Give up and pass responsibility on to the next filter
+                return null;
+            }
+
+            byte[] decodedHeader = Base64.getDecoder().decode(identityHeader);
+            Map authObject = mapper.readValue(decodedHeader, Map.class);
+            Map identity = (Map) authObject.getOrDefault("identity", Collections.emptyMap());
+            String accountNumber = (String) identity.get("account_number");
+
+            if (StringUtils.isEmpty(accountNumber)) {
+                throw new PreAuthenticatedCredentialsNotFoundException(RH_IDENTITY_HEADER +
+                    " contains no principal");
+            }
+
+            return accountNumber;
+        }
+        catch (IOException | PreAuthenticatedCredentialsNotFoundException e) {
+            throw new SubscriptionsException(ErrorCode.VALIDATION_FAILED_ERROR, Response.Status.BAD_REQUEST,
+                "error processing identity header", e);
```

Comment:
When this exception is thrown, it doesn't seem to get captured by the exception mappers. The correct exception appears in the logs, but I get the following response:

```
<!doctype html><html lang="en"><head><title>HTTP Status 500 – Internal Server Error</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 500 – Internal Server Error</h1></body></html>
```

Reply:
Ah, okay.  I think I know what's happening.  Spring Security tries to forward to an error page which is not useful here.  I'll have to fix this.